source/connection_monitor.c

```diff
@@ -0,0 +1,200 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/http/private/connection_monitor.h>
+
+#include <aws/http/connection.h>
+#include <aws/http/statistics.h>
+#include <aws/io/channel.h>
+#include <aws/io/logging.h>
+#include <aws/io/statistics.h>
+
+#include <aws/common/clock.h>
+
+#include <inttypes.h>
+
+static void s_process_statistics(
+    struct aws_crt_statistics_handler *handler,
+    struct aws_crt_statistics_sample_interval *interval,
+    struct aws_array_list *stats_list,
+    void *context) {
+
+    (void)interval;
+
+    struct aws_statistics_handler_http_connection_monitor_impl *impl = handler->impl;
+
+    uint64_t pending_read_interval_ns = 0;
+    uint64_t pending_write_interval_ns = 0;
+    uint64_t bytes_read = 0;
+    uint64_t bytes_written = 0;
+
+    /*
+     * Pull out the data needed to perform the throughput calculation
+     */
+    size_t stats_count = aws_array_list_length(stats_list);
+    for (size_t i = 0; i < stats_count; ++i) {
+        struct aws_crt_statistics_base *stats_base = NULL;
+        if (aws_array_list_get_at(stats_list, &stats_base, i)) {
+            continue;
+        }
+
+        switch (stats_base->category) {
+            case AWSCRT_STAT_CAT_SOCKET: {
+                struct aws_crt_statistics_socket *socket_stats = (struct aws_crt_statistics_socket *)stats_base;
+                bytes_read = socket_stats->bytes_read;
+                bytes_written = socket_stats->bytes_written;
+                break;
+            }
+
+            case AWSCRT_STAT_CAT_HTTP1: {
+                struct aws_crt_statistics_http1 *http1_stats = (struct aws_crt_statistics_http1 *)stats_base;
+                pending_read_interval_ns = http1_stats->pending_incoming_stream_ns;
+                pending_write_interval_ns = http1_stats->pending_outgoing_stream_ns;
+                break;
+            }
+
+            default:
+                break;
+        }
+    }
+
+    struct aws_channel *channel = context;
+
+    if (impl->options.minimum_throughput_bytes_per_second == 0 ||
+        impl->options.minimum_throughput_failure_threshold_in_seconds == 0) {
+        return;
+    }
+
+    /*
+     * All early-out/negative execution paths reset the failure count to zero.  Keep a copy of the current count for the
+     * remaining path.
+     */
+    uint32_t current_failure_count = impl->consecutive_throughput_failures;
+    impl->consecutive_throughput_failures = 0;
+
+    uint64_t pending_io_time_ns = 0;
+    if (aws_add_u64_checked(pending_read_interval_ns, pending_write_interval_ns, &pending_io_time_ns)) {
```

Comment:
I don't understand why you're adding together the read and write intervals. If there's a stream that's been sending the entire time, you'll end up with a `pending_io_time = 2 seconds`
and a `total_bytes = N sent bytes` which will get you `bytes_per_second = N/2`

I would think you'd want to report `bytes_per_second = N`

Maybe it would be better to calculate the `read_bytes_per_second` and `write_bytes_per_second` separately, and either ADD those or take the MAX of the two?

Reply:
That was just dumb on my part.